### PR TITLE
Add a monitoring provisioning flow

### PR DESCRIPTION
**Pull request by anmolbabu**
Add a monitoring provisioning flow

tendrl-bug-id: Tendrl/specifications#42
Signed-off-by: anmolbabu <anmolbudugutta@gmail.com>

### CODE REVIEW

**Comment by anmolbabu** [comment]
@shtripat @nthomas-redhat @r0h4n @brainfunked Please review

File: specs/configure_monitoring.adoc [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+== Introduction
+
+Tendrl's monitoring stack uses the following components:
+
+* collectd
+  ** A daemon process that resides on tendrl managed nodes, collects
+     performance statistics and pushes them to time-series db for maintaining
+     historic data.
+  ** Everything in collectd is done in plugins.
```
Comment (shtripat) [inline specs/configure_monitoring.adoc]: %s/Everything in collectd is done in plugins./Everything in collectd is achieved through plugins./g

File: specs/configure_monitoring.adoc [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+== Use Cases
+
+The use cases addressed by this spec are:
+
+* Tendrl core and monitoring stack are installed and user starts node-agent on
+  nodes so that they can be managed by tendrl and the new node needs to be
+  monitored.
+* Tendrl core stack is installed and one/more nodes are managed by tendrl core.
+  Now the user decides to monitor his tendrl managed nodes by installing the
+  tendrl/performance_monitoring.
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:44]: Another use case would be the user plugs out the monitoring after using for a while

File: specs/configure_monitoring.adoc [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+* The tendrl/performance_monitoring on start, uploads the default monitoring
+  configuration and flows for generating the collectd configurations.
+* The performance_monitoring application then looks into the etcd's '/nodes'
+  directory to see if there are any nodes that are currently managed by tendrl
+  If there are any, it triggers the monitoring provisioning flow on every node
+  for node monitoring via the core API.
+  ** The provisioning flow for the monitoring stack installs collectd-related
+     pieces on every node and generates the configuration files from templates
+     shipped with the collectd plugins on nodes.
+* For any new node that is managed by tendrl, the performance_monitoring
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:60]: How this is done? Please add the details

File: specs/configure_monitoring.adoc [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+    'swap': {'Warning': 50, 'Failure': 70}
+  }
+}
+----
+
+and they will be system wide default configuration that are exposed for
+updates in the form of a yaml file at /etc/tendrl/monitoring_defaults.yaml.
+These values are read at the time of starting of performance_monitoring
+application and hence updates to the file will be reflected on restart of
+performance monitoring application.
```
Comment (cybernth) [inline specs/configure_monitoring.adoc]: This configuration should be maintained in the central store. Provide APIs for update. Configuration on the central store should not be overwritten on restart.
Are we planning to provide cluster wide configuartion as well?

File: specs/configure_monitoring.adoc [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+=== Impacted Modules:
+
+==== Tendrl API impact:
+
+The flow defined for the configuration is intended to be intially triggered
+internally by the performance_monitoring applciation but for editting of the
+configuration (ex: updating the threshold values to override the defaults),
+the tendrl needs to proxy to the api exposed by the performance_monitoring
+application.
+
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:234]: Its good to provide the details here or is it provided some where else?

File: specs/configure_monitoring.adoc [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+proxies to.
+
+==== Tendrl/common impact:
+
+None
+
+==== Tendrl/node_agent impact:
+
+The flow framework needs to be reorganised as described in:
+https://github.com/Tendrl/specifications/pull/8
```
Comment (cybernth) [inline specs/configure_monitoring.adoc]: Any changes to the tendrl defintions? please capture those details here

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/configure_monitoring.adoc (hunk 3 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc:60]: Ok

File: specs/configure_monitoring.adoc (hunk 2 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc:44]: Ok

File: specs/configure_monitoring.adoc (hunk 4 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc]: >This configuration should be maintained in the central store. Provide APIs for update.

Ok

> Configuration on the central >store should not be overwritten on restart.

Yes

> Are we planning to provide cluster wide configuartion as well?

Yes I'll add more details on this.

File: specs/configure_monitoring.adoc (hunk 5 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc:234]: By details did you mean the api formats and parameters. I'll add it as part of this patch

File: specs/configure_monitoring.adoc (hunk 6 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc]: No changes in definitions but just that the path to which these definitions will be sent and the way it'll be fetched will change slightly.

**Review by r0h4n — APPROVED** [review]
@Tendrl/tendrl-core

File: specs/configure_monitoring.adoc [hunk 7]
```diff
@@ -0,0 +1,304 @@
...
+  monitored.
+* Tendrl core stack is installed and one/more nodes are managed by tendrl core.
+  Now the user decides to monitor his tendrl managed nodes by installing the
+  tendrl/performance_monitoring.
+* User plugs out the monitoring after using for a while.
+
+== Proposed change
+
+* The tendrl/performance_monitoring on start, checks the existence of
+  '/monitoring/defaults' directory in etcd and if its not present, it uploads
```
Comment (r0h4n) [inline specs/configure_monitoring.adoc:50]: make use of a "definition" object in the Namespace of tendrl.performance_monitoring and follow as per 
https://github.com/Tendrl/specifications/issues/32

File: specs/configure_monitoring.adoc [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+  the default monitoring configuration and flows for generating the collectd
+  configurations.
+* The performance_monitoring application then looks into the etcd's '/nodes'
+  directory to see if there are any nodes that are currently managed by tendrl.
+  If there are any, it triggers the monitoring provisioning flow on every node
+  for node monitoring via the core API.
+  ** The provisioning flow for the monitoring stack installs collectd-related
```
Comment (r0h4n) [inline specs/configure_monitoring.adoc:57]: Need the flow definition

**Review by r0h4n — CHANGES_REQUESTED** [review]

**Comment by r0h4n** [comment]
flow definition required, rest looks good

File: specs/configure_monitoring.adoc (hunk 7 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc:50]: @r0h4n Why this needs to be part of a flow definition? This is something that is supposed to happen at the start of the application(not triggered by user explicitly). Also, could you please add sample definition in the specs addressing #32

File: specs/configure_monitoring.adoc (hunk 8 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc:57]: Ok

**Comment by anmolbabu** [comment]
@Tendrl/tendrl-core Please review

**Comment by anmolbabu** [comment]
@r0h4n @nthomas-redhat @shtripat I have incorporated your comments. Kindly review and provide your valuable feedback

File: specs/configure_monitoring.adoc [hunk 9]
```diff
@@ -162,17 +290,11 @@ https://github.com/Tendrl/specifications/pull/8/files
...
 == Documentation impact:
 
-The items to be documented includes an api exposed for overriding the default
-monitoring configuration and also a config file that can be used to have custom
-default values that override the tendrl chosen defaults as desscribed under:
-"Other end user impact" and the api is described in "Tendrl API impact"
-
+This adds apis as described in "Tendrl API impact" which need to be documented
 
 == References:
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:413]: will the tendrl-api related changes captured ion a different spec? I think it shouldnot. You need to raise github issues to track those and add the links here

File: specs/configure_monitoring.adoc [hunk 10]
```diff
@@ -42,22 +42,159 @@ The use cases addressed by this spec are:
...
+  existence
+* This spec requires implementation of the following:
+  ** The api to modify defaults which will only modify default configuration
+  values on the central store and will only be picked up for fresh
+  configurations and hence already configured nodes will not be affected by
+  modifications in default configurations. The api as exposed by the
+  performance-monitoring will look like:
 
+----
+
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:80]: The APIs needs to be reviewed by @anivargi / @brainfunked

File: specs/configure_monitoring.adoc [hunk 11]
```diff
@@ -42,22 +42,159 @@ The use cases addressed by this spec are:
...
+  on that node.This api then loads a job with the api parameters as job
+  parameters along with the fqdn of the new node. The job is then executed by
+  node-agent residing on that node.
+* If the user decides not to have monitoring anymore, an api will be exposed
+  by the core API which will internally invoke a flow that does the following:
+  **  Stop collectd on every tendrl managed node
+  **  Stop performance-monitoring application
+  This means that even though the monitoring is turned off, the actual
+  monitoring related bits are only terminated from service and not removed from
+  existence
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:71]: you mean packages remain installed?

File: specs/configure_monitoring.adoc [hunk 12]
```diff
@@ -235,15 +235,93 @@ This spec requires 2 kinds of new apis to be exposed:
...
 
 ==== Tendrl/node_agent impact:
 
-The flow framework needs to be reorganised as described in:
-https://github.com/Tendrl/specifications/pull/8
+A new atom 'check_service_status' is implemented and a corresponding entry is
+made under 'Service' object of node-agent definition file as:
+
+----
+    Service:
```
Comment (cybernth) [inline specs/configure_monitoring.adoc:317]: I think this should go into common name space. @brainfunked , @r0h4n

File: specs/configure_monitoring.adoc (hunk 11 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc:71]: Yes

File: specs/configure_monitoring.adoc (hunk 12 above)
Comment (cybernth) [reply specs/configure_monitoring.adoc:317]: @anmolbabu , yes this service related stuff should be moved to common name space

Comment (anmolbabu) [reply specs/configure_monitoring.adoc:317]: Ok

File: specs/configure_monitoring.adoc (hunk 5 above)
Comment (cybernth) [reply specs/configure_monitoring.adoc:234]: yes

**Review by cybernth — COMMENTED** [review]
Looks good to me. I will approve if the APIs are reviewed by @anivargi

**Comment by cybernth** [comment]
@Tendrl/qe , please have a look

**Comment by anmolbabu** [comment]
@anivargi please review

**Review by shtripat — COMMENTED** [review]
Looks good to me. Anup to review api changes

**Review by anivargi — APPROVED** [review]

File: specs/configure_monitoring.adoc (hunk 7 above)
Comment (r0h4n) [reply specs/configure_monitoring.adoc:50]: https://github.com/Tendrl/specifications/blob/master/specs/compile_and_centralize_tendrl_definitions_access.adoc#proposed-change

**Review by r0h4n — CHANGES_REQUESTED** [review]
Please put integration config and definitions via "Config" and "Tendrl_Definition" objects 

https://github.com/Tendrl/specifications/blob/master/specs/compile_and_centralize_tendrl_definitions_access.adoc#proposed-change

**Review by cybernth — APPROVED** [review]

**Comment by anmolbabu** [comment]
@r0h4n 

> Please put integration config and definitions via "Config" and "Tendrl_Definition" objects
>https://github.com/Tendrl/specifications/blob/master/specs/compile_and_centralize_tendrl_definitions_access.adoc#proposed-change

I tried to follow the link above, but I didn't find a sample flow implementing this and also the purpose served by 'Tendrl_Definition' and 'Config' objects. I think you might have given me a different link by mistake. I have updated my patch set in accordance with my understanding. Kindly review and correct me if I am wrong.

**Review by shtripat — COMMENTED** [review]
LGTM. @r0h4n please check if your requested changes are fine.

**Comment by r0h4n** [comment]
@anmolbabu 

- Fix central store location for definitions under node context at "/_tendrl/definitions/master".
- Fix central store location for definitions which extend the node_agent (eg: Monitoring, alerts) at "/_tendrl/definitions/integrations.d/<repo_name>/data" via a defined object "Tendrl_definition" in the Monitoring, Alerts namespace.

- Then these definitions will be included in the node agent master yaml via "include: _tendrl/definitions/integrations.d/*"

- "Config" object is for pushing component config to "_tendrl/config/<repo_name>"

Does that make sense? You simply have to define these in your tendrl namespace as any other object

File: specs/configure_monitoring.adoc [hunk 13]
```diff
@@ -240,35 +240,27 @@ will look like:
...
+          - plugin_conf_params
+          - Service.name
+      post_run:
+        - tendrl.node_agent.objects.Service.atoms.check_service_status
+      pre_run:
+        - tendrl.node_agent.objects.Node.atoms.check_node_up
+      run: tendrl.node_monitoring.flows.configure_collectd.ConfigureCollectd
+      type: Create
+      uuid: dc8fff3a-34d9-4786-9282-55eff6abb6c4
+  Tendrl_definition: '/_tendrl/definitions/integrations.d/performance_monitoring/data'
```
Comment (r0h4n) [inline specs/configure_monitoring.adoc]: These do not confirm to Tendrl Definitions schema,  Tendrl_definition and Config are "objects"

**Review by r0h4n — CHANGES_REQUESTED** [review]

File: specs/configure_monitoring.adoc (hunk 13 above)
Comment (anmolbabu) [reply specs/configure_monitoring.adoc]: Ok

File: specs/configure_monitoring.adoc [hunk 14]
```diff
@@ -240,6 +240,28 @@ will look like:
 
 ----
 namespace.tendrl.node_monitoring:
+  objects:
+    Tendrl_definition:
+      attrs:
+        data:
+          help: "The flow definitions corresponding to performance_monitoring module"
+          type: json
+        value:
```
Comment (r0h4n) [inline specs/configure_monitoring.adoc]: looks good, I missed mentioning in last review, this is how the yaml should look

> namespace.tendrl.node_monitoring:
>   objects:
>     Tendrl_definition:
>       value:/_tendrl/definitions/integrations.d/performance_monitoring/data
>       attrs:
>         data:
>           help: "The flow definitions corresponding to performance_monitoring module"
>           type: json
>     Config:
>         value: _tendrl/config/performance_monitoring
>         attrs:
>         data:
>           help: "The flow definitions corresponding to performance_monitoring module"
>           type: json
> tendrl_schema_version: 0.3
>

**Review by r0h4n — CHANGES_REQUESTED** [review]

**Review by r0h4n — APPROVED** [review]

**Comment by r0h4n** [comment]
@Tendrl/qe @Tendrl/tendrl-core @Tendrl/tendrl_api Please approve

**Review by cybernth — APPROVED** [review]
Latest changes are based on - https://github.com/Tendrl/specifications/pull/17 - looks good to me

File: specs/configure_monitoring.adoc (hunk 11 above)
Comment (brainfunked) [reply specs/configure_monitoring.adoc:71]: I think we can de-prioritise the turn-off mechanism for now and concentrate on the installation and use of the monitoring stack. I would rather not have the turn-off mechanism take any time away for now.

**Review by brainfunked — COMMENTED** [review]
I think we should expand the scope of this specification to include details about how the monitoring stack can be provisioned via the node agent. This needs a separate specification that I'll work on, as it would have to tie together other specifications such as the import cluster workflow, https://github.com/Tendrl/specifications/issues/54. We can put this spec on hold for a couple of days.

File: specs/configure_monitoring.adoc [hunk 15]
```diff
@@ -0,0 +1,418 @@
...
+    *** graphite-web - Graphite's user interface & API for rendering graphs
+        and dashboards
+  ** Metrics get fed into the stack via the Carbon service, which writes data
+     out to Whisper databases for long-term storage.
+
+== Problem Description
+
+Tendrl needs to configure collectd for it to start collecting the performance
+statistics at regular intervals of time. This process of configuring collectd
+should require no/minimal user intervention.
```
Comment (mbukatov) [inline specs/configure_monitoring.adoc:33]: Don't we need to explicitly state that we are solving historic time-series data problem here as opposed to reporting that something just happened immediately (which requires to deliver such message in given time range, and which is not part of this specification)?

I would like to make this clear immediately in the description (if this is the case).

**Comment by r0h4n** [comment]
This is completed, let me know if this needs to be re-opened

**Pull request merged**